### PR TITLE
BENCH: optimize: milp: add MILP benchmarks

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-15460 

#### What does this implement/fix?
A dedicated function for mixed integer linear programming was added in gh-15460. This PR adds benchmarks to track its performance over time.

### CODE REVIEW

**Comment by mdhaber** [comment]
@mckib2 Benchmark failure is after these new benchmarks ran:
```
[ 20.79%] ··· Running (optimize_linprog.MagicSquare.time_magic_square--)....
[ 22.04%] ··· Running (optimize_milp.MilpMiplibBenchmarks.time_milp--).
[ 22.22%] ··· Running (optimize_zeros.Newton.time_newton--).............
```
On my machine, the results of `asv run --bench optimize_milp` are:
```
(scipydev) C:\Users\matth\scipy\benchmarks>asv run --bench optimize_milp
· No `environment_type` specified in asv.conf.json. This will be required in the future.
· Creating environments
· Discovering benchmarks
· Running 2 total benchmarks (1 commits * 1 environments * 2 benchmarks)
[  0.00%] · For scipy commit 1b6173f9 <milp_benchmarks>:
[  0.00%] ·· Benchmarking conda-py3.9-Cython0.29.21-numpy-pybind11-pytest-pythran
[ 25.00%] ··· Running (optimize_milp.MilpMagicSquare.time_magic_square--).
[ 50.00%] ··· Running (optimize_milp.MilpMiplibBenchmarks.time_milp--).
[ 75.00%] ··· optimize_milp.MilpMagicSquare.time_magic_square                                                        ok
[ 75.00%] ··· ====== =============
               size
              ------ -------------
                3     10.5±0.09ms
                4       897±6ms
                5      3.01±0.01s
                6      22.6±0.06s
              ====== =============

[100.00%] ··· optimize_milp.MilpMiplibBenchmarks.time_milp                                                           ok
[100.00%] ··· ============= ===========
                 problem
              ------------- -----------
               piperout-27   25.3±0.3s
              ============= ===========
```
If the size 6 magic square and piperout problem are too long, I suppose we can add an xslow skip mechanism.
Otherwise, I think the lint failure can be fixed without re-running CI, and then this can be merged.

File: benchmarks/benchmarks/optimize_milp.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+with safe_import():
+    from scipy.optimize import milp
+
+with safe_import():
+    from scipy.optimize.tests.test_linprog import magic_square
+
+
+# MIPLIB 2017 benchmarks included with permission of the authors
+milp_problems = ["piperout-27"]
```
Comment (rgommers) [inline benchmarks/benchmarks/optimize_milp.py:16]: I don't see a script or explanation for how the `.npz` file in this PR was created. That would be useful to add.

File: benchmarks/benchmarks/optimize_milp.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    params = [[3, 4, 5, 6]]
+    param_names = ['size']
+
+    def setup(self, n):
+        A_eq, b_eq, self.c, self.numbers, self.M = magic_square(n)
+        self.constraints = (A_eq, b_eq, b_eq)
+
+    def time_magic_square(self, n):
+        res = milp(c=self.c*0, constraints=self.constraints,
+                      bounds=(0, 1), integrality=True)
```
Comment (mckib2) [inline benchmarks/benchmarks/optimize_milp.py:63]: Nitpick: indentation

Comment (mdhaber) [reply benchmarks/benchmarks/optimize_milp.py:63]:
```suggestion
                   bounds=(0, 1), integrality=True)
```

File: benchmarks/benchmarks/optimize_milp.py (hunk 1 above)
Comment (mdhaber) [reply benchmarks/benchmarks/optimize_milp.py:16]: @mckib2 converted the problem from `mps` format to arrays using his `scikit-glpk`, then we saved them to `.npz`. Do you want that stated here? e.g.:
> Loaded from original MPS format using scikit-glpk

Comment (mckib2) [reply benchmarks/benchmarks/optimize_milp.py:16]: I suspect if anything needs to be done here, it's just a hint that an (any) MPS converter was used to load the problem into Python and then rearranged in the linprog-style and saved back out to `npz`

**Comment by mckib2** [comment]
> @mckib2 Benchmark failure is after these new benchmarks ran:

As long as it's not indicating a crash, I think it's fine to mark as xslow (many MIP problems are!)

**Comment by mdhaber** [comment]
> As long as it's not indicating a crash

No, what I meant is that these benchmarks ran fine. ASV was timing out in main until recently, but I was noting that despite that, we were able to see that the benchmarks ran OK in CI.

File: benchmarks/benchmarks/optimize_milp.py (hunk 1 above)
Comment (mdhaber) [reply benchmarks/benchmarks/optimize_milp.py:16]:
```suggestion
# The MIPLIB benchmark problem set was downloaded from https://miplib.zib.de/
# An MPS converter (scikit-glpk) was used to load the problem into Python. The
# arrays were arranged to the format required by `milp` and saved to `npz`
# format using `np.savez`.
milp_problems = ["piperout-27"]
```

File: benchmarks/benchmarks/optimize_milp.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        if not hasattr(self, 'data'):
+            dir_path = os.path.dirname(os.path.realpath(__file__))
+            datafile = os.path.join(dir_path, "linprog_benchmark_files",
+                                    "milp_benchmarks.npz")
+            self.data = np.load(datafile, allow_pickle=True)
+
+        c, A_ub, b_ub, A_eq, b_eq, bounds, integrality = self.data[prob]
+
+        lb = [l for l, u in bounds]
+        ub = [u for l, u in bounds]
```
Comment (mdhaber) [inline benchmarks/benchmarks/optimize_milp.py:33]:
```suggestion
        lb = [li for li, ui in bounds]
        ub = [ui for li, ui in bounds]
```

File: benchmarks/benchmarks/optimize_milp.py (hunk 1 above)
Comment (mdhaber) [reply benchmarks/benchmarks/optimize_milp.py:20]: Hinted.

**Comment by mdhaber** [comment]
OK @mckib2 If this looks good to you, it's ready from my perspective. The benchmarks timeout persists because I didn't merge `main`, but again, the `milp` benchmarks ran successfully already:
```
[ 20.79%] ··· Running (optimize_linprog.MagicSquare.time_magic_square--)....
[ 22.04%] ··· Running (optimize_milp.MilpMiplibBenchmarks.time_milp--).
[ 22.22%] ··· Running (optimize_zeros.Newton.time_newton--).............
```

File: benchmarks/benchmarks/optimize_milp.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+# arrays were arranged to the format required by `milp` and saved to `npz`
+# format using `np.savez`.
+milp_problems = ["piperout-27"]
+
+
+class MilpMiplibBenchmarks(Benchmark):
+    params = [milp_problems]
+    param_names = ['problem']
+
+    def setup(self, prob):
```
Comment (tylerjereddy) [inline benchmarks/benchmarks/optimize_milp.py:27]: Here and in the other benchmark, is it really better to have the parameter name different from the name in `param_names`?

Maybe it is just habit for me, but I usually just use the same names that I put in the `param_names` list, which also show up in the benchmark output I think.

File: benchmarks/benchmarks/optimize_milp.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        res = milp(c=self.c*0, constraints=self.constraints,
+                   bounds=(0, 1), integrality=True)
+        assert res.status == 0
+        x = np.round(res.x)
+        s = (self.numbers.flatten() * x).reshape(n**2, n, n)
+        square = np.sum(s, axis=0)
+        assert_allclose(square.sum(axis=0), self.M)
+        assert_allclose(square.sum(axis=1), self.M)
+        assert_allclose(np.diag(square).sum(), self.M)
+        assert_allclose(np.diag(square[:, ::-1]).sum(), self.M)
```
Comment (tylerjereddy) [inline benchmarks/benchmarks/optimize_milp.py:75]: Is the numerical behavior of the code so unstable that we can't keep the behavior under test in the main test suite and have the benchmarks do timing only?

File: benchmarks/benchmarks/optimize_milp.py (hunk 4 above)
Comment (mdhaber) [reply benchmarks/benchmarks/optimize_milp.py:27]: No. I prepared this by copy-pasting and not paying too much attention to it.

File: benchmarks/benchmarks/optimize_milp.py (hunk 5 above)
Comment (mdhaber) [reply benchmarks/benchmarks/optimize_milp.py:75]: Not so unstable. This mostly shows that the benchmark is solving the problem it intends to solve.

**Comment by mckib2** [comment]
Everything looks good to me, will be good to get these benchmarks in.  Thanks @mdhaber !

**Pull request merged**